src/librustc_middle/mir/interpret/queries.rs

```diff
@@ -69,9 +69,29 @@ impl<'tcx> TyCtxt<'tcx> {
         // improve caching of queries.
         let inputs = self.erase_regions(&param_env.and(cid));
         if let Some(span) = span {
-            self.at(span).const_eval_validated(inputs)
+            self.at(span).const_eval_for_ty(inputs)
         } else {
-            self.const_eval_validated(inputs)
+            self.const_eval_for_ty(inputs)
         }
     }
+
+    /// Evaluate a static's initializer, returning the allocation of the initializer's memory.
+    pub fn eval_static_initializer(self, def_id: DefId) -> Result<&'tcx Allocation, ErrorHandled> {
+        trace!("eval_static_initializer: Need to compute {:?}", def_id);
+        assert!(self.is_static(def_id));
+        let instance = ty::Instance::mono(self, def_id);
+        let gid = GlobalId { instance, promoted: None };
+        self.eval_mir_const(gid, ty::ParamEnv::reveal_all())
+    }
+
+    /// Evaluate anything constant-like, returning the allocation of the final memory.
+    fn eval_mir_const(
```

Comment:
Since `mir::Const` will likely be something different, this name is still potentially confusing,

An alternative would be `eval_global`. But that is very close to `const_eval_global_id`, which maybe should be `eval_global_for_ty`.

Why do some functions that with `const_eval` and some with `eval`?

Reply:
imo we should phase out `const_eval_*` in favour of just `eval_*`

how about `eval_to_allocation`

Reply:
> `eval_to_allocation`

:+1:
The query should probably be called `const_eval_to_allocation` then?

Reply:
there's no evaluation other than constant evaluation in rustc. We already use just `eval` for the method name on constants. We can keep using the `raw` postfix just like `is_freeze` and `is_freeze_raw` do.

Reply:
Or `eval_to_allocation` for the query, whatever.

Oh but that's a different "raw" than before where "raw" meant "unvalidated"? If this is just meant to indicate "query" IMO `eval_to_allocation_query` would be a better name but if we already have a pattern for this then I don't really care strongly.

Reply:
If this query is `eval_to_allocation`, then the other underlying query should probably be called `eval_to_valtree` or so? At least once we have valtrees.

Also all these helpers like `const_eval_global_id` here do not indicate whether they return a valtree or an allocation.

Reply:
yea, I gotta analyze all of the uses of these helpers and split them where appropriate.

> then the other underlying query should probably be called `eval_to_valtree_raw` or so? At least once we have valtrees.

yes. We can name it `eval_to_const_value` until we have valtrees.

Reply:
Makes sense.

Reply:
> yea, I gotta analyze all of the uses of these helpers and split them where appropriate.

That can be a separate PR, as far as I am concerned.